### PR TITLE
Gallery List Demo

### LINKED ISSUE #1573: Gallery App has Lists
From https://www.google.com/design/spec/components/lists.html#

- [ ] scrolling, vertical
- [ ] sortable
- [ ] content: action (image), text, secondary action (smaller image)
- [ ] content: text-only
- [ ] content: avatar + text
- [ ] one-line text
- [ ] two-line text
- [ ] three-line text

### PR DESCRIPTION
**Pull request by HansMuller**
Added a List demo that illustrates the various standard list item configurations, scrolling, and sorting.

ListItem Changes:
- Center is now two widgets called "primary" and "secondary" per the spec.
- The item's height depends the presence of the secondary widget and an isThreeLine flag. If isThreeLine is true, the secondary widget is allocated enough space for two lines.
- The left widget is vertically centered for 1 or 2 line items, top justified for 3 lines. It's always left justified.
- The right widget is vertically centered for 1 or 2 line items, top justified for 3 lines. It's always right justified.
- Added support for dense layouts.

Also:
- CircleAvatar now has a Widget child, instead of a string label.
- I made a small change (OK, it's a hack) in Scaffold to get it to rebuild the persistent bottom sheet along with everything else. Without the change, app.setState() doesn't rebuild the bottom sheet. Ideally, the bottom sheet would be rebuilt by the Scaffold directly. I was too weak to sort out how to do that given showBottomSheet(), ScaffoldFeatureController, etc.

Fixes https://github.com/flutter/flutter/issues/1573

### CODE REVIEW

File: packages/flutter/lib/src/material/circle_avatar.dart [hunk 1]
```diff
@@ -11,35 +11,32 @@ import 'typography.dart';
 class CircleAvatar extends StatelessComponent {
   CircleAvatar({
     Key key,
-    this.label,
+    this.child,
```
Comment (Hixie) [inline packages/flutter/lib/src/material/circle_avatar.dart]: why this change? It doesn't look like you use it to do anything but text yet.

File: examples/material_gallery/lib/demo/list_demo.dart [hunk 2]
```diff
@@ -0,0 +1,195 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+
+import 'package:flutter/rendering.dart';  // TBD remove this
```
Comment (Hixie) [inline examples/material_gallery/lib/demo/list_demo.dart]: remove this

File: examples/material_gallery/lib/demo/list_demo.dart [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+enum ListDemoItemSize {
+  oneLine,
+  twoLine,
+  threeLine
+}
+
+class ListDemo extends StatefulComponent {
+  ListDemoState createState() => new ListDemoState();
```
Comment (Hixie) [inline examples/material_gallery/lib/demo/list_demo.dart]: constructor should take a key

File: examples/material_gallery/lib/demo/list_demo.dart [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+  List<String> items = [];
+  ListDemoItemSize itemSize = ListDemoItemSize.threeLine;
+  bool isDense = true;
+  bool showAvatar = true;
+  bool showIcon = false;
+  bool reverseSort = false;
+
+  void initState() {
+    super.initState();
+    items = <String>[
```
Comment (Hixie) [inline examples/material_gallery/lib/demo/list_demo.dart]: why not do this in the initialiser?

File: examples/material_gallery/lib/demo/list_demo.dart [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+  threeLine
+}
+
+class ListDemo extends StatefulComponent {
+  ListDemoState createState() => new ListDemoState();
+}
+
+class ListDemoState extends State<ListDemo> {
+  final GlobalKey<ScaffoldState> scaffoldKey = new GlobalKey<ScaffoldState>();
+  List<String> items = [];
```
Comment (Hixie) [inline examples/material_gallery/lib/demo/list_demo.dart]: type

File: examples/material_gallery/lib/demo/list_demo.dart [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+  ListDemoItemSize itemSize = ListDemoItemSize.threeLine;
+  bool isDense = true;
+  bool showAvatar = true;
+  bool showIcon = false;
+  bool reverseSort = false;
+
```
Comment (Hixie) [inline examples/material_gallery/lib/demo/list_demo.dart]: all these state variables should be privates

File: examples/material_gallery/lib/demo/list_demo.dart [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+  }
+
+  void changeItemSize(ListDemoItemSize size) {
+    setState(() {
+      itemSize = size;
+    });
+  }
+
+  void showConfigurationSheet(BuildContext context) {
+    scaffoldKey.currentState.showBottomSheet((_) {
```
Comment (Hixie) [inline examples/material_gallery/lib/demo/list_demo.dart]: please don't use `_` arguments in sample code

File: examples/material_gallery/lib/demo/list_demo.dart [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+              right: new Checkbox(
+                value: showIcon,
+                onChanged: (bool value) {
+                  setState(() {
+                    showIcon = value;
+                  });
+                }
+              )
+            ),
+            new ListItem(
```
Comment (Hixie) [inline examples/material_gallery/lib/demo/list_demo.dart]: we really should make tapping the list item also tap the widget

File: packages/flutter/lib/src/material/circle_avatar.dart (hunk 1 above)
Comment (HansMuller) [inline packages/flutter/lib/src/material/circle_avatar.dart]: It's to allow for the common case of a circular Avatar with an image inside.

File: examples/material_gallery/lib/demo/list_demo.dart [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+
+  Widget buildListItem(BuildContext context, String item) {
+    Widget secondary;
+    if (itemSize == ListDemoItemSize.twoLine) {
+      secondary = new Text(
+        "Additional item information."
+      );
+    } else if (itemSize == ListDemoItemSize.threeLine) {
+      secondary = new Text(
+        "Even more additional list item information appears on line three. "
```
Comment (Hixie) [inline examples/material_gallery/lib/demo/list_demo.dart]: trailing space

File: examples/material_gallery/lib/demo/list_demo.dart (hunk 2 above)
Comment (HansMuller) [inline examples/material_gallery/lib/demo/list_demo.dart]: Just telling myself wasn't good enough. Oops.

File: examples/material_gallery/lib/demo/list_demo.dart [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+      isDense: isDense,
+      left: showAvatar ? new CircleAvatar(child: new Text(item)) : null,
+      primary: new Text('This item represents $item'),
+      secondary: secondary,
+      right: showIcon ? new Icon(icon: 'action/info', color: Theme.of(context).disabledColor) : null
+    );
+  }
+
+  Widget build(BuildContext context) {
+    final String layoutText = isDense ? " - Dense" : "";
```
Comment (Hixie) [inline examples/material_gallery/lib/demo/list_demo.dart]: Use `\u2013` instead of `-` for bonus typography points

File: examples/material_gallery/lib/demo/list_demo.dart [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+        itemSizeText = 'Two-Line List';
+        break;
+      case ListDemoItemSize.threeLine:
+        itemSizeText = 'Three-Line List';
+        break;
+    }
+    return new Scaffold(
+      key: scaffoldKey,
+      toolBar: new ToolBar(
+        center: new Text('${itemSizeText}${layoutText}'),
```
Comment (Hixie) [inline examples/material_gallery/lib/demo/list_demo.dart]: travis will complain about the braces here

File: examples/material_gallery/lib/demo/list_demo.dart [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+            tooltip: 'Sort',
+            onPressed: () {
+              setState(() {
+                reverseSort = !reverseSort;
+                items.sort((String a, String b) => reverseSort ? a.compareTo(b) : b.compareTo(a));
+              });
+            }
+          ),
+          new IconButton(
+            icon: "navigation/more_vert",
```
Comment (Hixie) [inline examples/material_gallery/lib/demo/list_demo.dart]: I don't think this icon should be used to show a bottom sheet. Maybe action/settings?

File: examples/material_gallery/lib/demo/list_demo.dart [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+        child: new Block(
+          children: items.map((String item) => buildListItem(context, item)).toList()
+        )
+      )
+    );
+  }
+}
+
+void main() {
+  //debugPaintSizeEnabled = true;
```
Comment (Hixie) [inline examples/material_gallery/lib/demo/list_demo.dart]: commented out code

File: examples/material_gallery/lib/demo/list_demo.dart (hunk 4 above)
Comment (HansMuller) [inline examples/material_gallery/lib/demo/list_demo.dart]: I assume that there's some tradeoff about work done when the class is loaded vs when it's instantiated the first time. In this case I was not considering any tradeoffs :-). I've moved the value to the initializer.

File: packages/flutter/lib/src/material/scaffold.dart [hunk 14]
```diff
@@ -484,8 +484,10 @@ class ScaffoldState extends State<Scaffold> {
       List<Widget> bottomSheets = <Widget>[];
       if (_dismissedBottomSheets != null && _dismissedBottomSheets.isNotEmpty)
         bottomSheets.addAll(_dismissedBottomSheets);
-      if (_currentBottomSheet != null)
+      if (_currentBottomSheet != null) {
+        _currentBottomSheet.setState(() { });
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: noooooooo, don't call setState from build.

File: packages/flutter/lib/src/material/scaffold.dart [hunk 15]
```diff
@@ -332,7 +332,7 @@ class ScaffoldState extends State<Scaffold> {
         bottomSheet,
         completer,
         () => entry.remove(),
-        setState
+        (VoidCallback fn) { bottomSheetKey.currentState?.setState(fn); }
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: this is fine. I'd even add an assert that bottomSheetKey.currentState isn't null.

File: examples/material_gallery/lib/demo/list_demo.dart (hunk 12 above)
Comment (HansMuller) [inline examples/material_gallery/lib/demo/list_demo.dart]: I lifted this from the original mocks for the Material Design gallery.

File: packages/flutter/lib/src/material/scaffold.dart (hunk 14 above)
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: I think the right thing to do here is to have `_PersistentBottomSheet` have a `clone()` method that returns a new identical widget, and put that one into the bottomSheets list instead of _currentBottomSheet._widget.

Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: but maybe @abarth has an idea. The problem here is that the scaffold wants to make sure that its bottom sheet's builder is invoked whenever it is itself invoked, but in between the scaffold and the builder there's another widget whose description we've pre-built and keep inserting (so it terminates the build walk and the builder never gets called).

File: examples/material_gallery/lib/demo/list_demo.dart (hunk 12 above)
Comment (Hixie) [inline examples/material_gallery/lib/demo/list_demo.dart]: That's very dubious. I'd change it to action/settings anyway. :-)

File: examples/material_gallery/lib/demo/list_demo.dart (hunk 4 above)
Comment (Hixie) [inline examples/material_gallery/lib/demo/list_demo.dart]: I think the performance here would be identical, it's just nice to be able to avoid the initState if you can.

File: packages/flutter/lib/src/material/circle_avatar.dart (hunk 1 above)
Comment (Hixie) [inline packages/flutter/lib/src/material/circle_avatar.dart]: I figured, I'm just not sure this is the right way to do it. It makes the code a lot more verbose in the text case. I'd wait until we have an image case before trying to solve that problem.

**Comment by Hixie** [comment]
lgtm modulo comments and travis.

Can you add a test for the bottom sheet statefulness thing?

File: examples/material_gallery/lib/demo/list_demo.dart (hunk 8 above)
Comment (HansMuller) [inline examples/material_gallery/lib/demo/list_demo.dart]: I filed an issue:
https://github.com/flutter/flutter/issues/2111

File: packages/flutter/lib/src/material/scaffold.dart (hunk 15 above)
Comment (HansMuller) [inline packages/flutter/lib/src/material/scaffold.dart]: There don't seem top be any uses of ScaffoldFeatureController.setState() except for the hack I added (and will remove).  However: I suppose the idea is that someone who's only holding on to the feature controller can update some state that the feature controller depends inside of myFeatureController.setState().

That's what I've done; it does work. It's not as tidy as just calling setState() once, on the application itself.

File: packages/flutter/lib/src/material/scaffold.dart (hunk 14 above)
Comment (HansMuller) [inline packages/flutter/lib/src/material/scaffold.dart]: I've removed this setState() call. Will file an issue about making this scenario work.

File: packages/flutter/lib/src/material/scaffold.dart (hunk 15 above)
Comment (HansMuller) [inline packages/flutter/lib/src/material/scaffold.dart]: I've filed a related issue: Rebuilding the Scaffold doesn't rebuild its BottomSheet
https://github.com/flutter/flutter/issues/2115

**Comment by HansMuller** [comment]
If you don't mind, I will add a test that confirms that ScaffoldFeatureController.setState for a BottomSheet does cause the BottomSheet to be rebuilt in a separate CL.

File: packages/flutter/lib/src/material/list_item.dart [hunk 16]
```diff
@@ -5,53 +5,123 @@
...
+/// Icons are defined with the [left] and [right] parameters. The first line of text
+/// is not optional and is specified with [primary]. The value of [secondary] will
+/// occupy the space allocated for an aditional line of text, or two lines if
+/// isThreeLine: true is specified. If isDense: true is specified then the overall
+/// height of this list item and the size of the DefaultTextStyles that wrap
+/// the [primary] and [secondary] widget are reduced.
 class ListItem extends StatelessComponent {
   ListItem({
     Key key,
     this.left,
```
Comment (abarth) [inline packages/flutter/lib/src/material/list_item.dart]: We might want to rename these left/right properties because they'll be reversed in RTL layouts.  Maybe "avatar" and "action" ?

File: packages/flutter/lib/src/material/list_item.dart [hunk 17]
```diff
@@ -5,53 +5,123 @@
...
-  final Widget center;
+  final Widget primary;
+  final Widget secondary;
   final Widget right;
+  final bool isThreeLine;
+  final bool isDense;
   final GestureTapCallback onTap;
   final GestureLongPressCallback onLongPress;
 
+  TextStyle primaryTextStyle(BuildContext context) {
```
Comment (abarth) [inline packages/flutter/lib/src/material/list_item.dart]: Should this be private?

File: packages/flutter/lib/src/material/list_item.dart [hunk 18]
```diff
@@ -5,53 +5,123 @@
...
+    final TextStyle style = Theme.of(context).text.subhead;
+    return isDense ? style.copyWith(fontSize: 13.0) : style;
+  }
+
+  TextStyle secondaryTextStyle(BuildContext context) {
```
Comment (abarth) [inline packages/flutter/lib/src/material/list_item.dart]: ditto

File: packages/flutter/lib/src/material/list_item.dart (hunk 16 above)
Comment (HansMuller) [inline packages/flutter/lib/src/material/list_item.dart]: I suppose that when we add support for RTL layouts there will be many left and rights to contend with. Toolbar, margins, padding, Positioned, etc.

Unfortunately the MD spec doesn't provide a unique name for the left and right sides. There's this:
primary
and this:
left

I suppose that we can refer to the left and right widgets however we want. Devs will have to learn what they mean.

File: packages/flutter/lib/src/material/list_item.dart (hunk 17 above)
Comment (HansMuller) [inline packages/flutter/lib/src/material/list_item.dart]: I did this so that it would be easy to override the primary text style in a subclass.  Like:
```
MyListItem extends ListItem {
  TextStyle primaryTextStyle(BuildContext context) => super.primaryTextStyle(context).copyWith(...);
}
```

Comment (abarth) [inline packages/flutter/lib/src/material/list_item.dart]: Ok.
